File: curriculum/13_api_gateway_circuit_breaker/node-impl/src/bulkhead.ts
export interface BulkheadSnapshot {
  maxConcurrency: number;
  inFlight: number;
  rejections: number;
}

export class Bulkhead {
  private inFlight = 0;
  private rejections = 0;

  constructor(private max: number) {}

  acquire(): boolean {
    if (this.inFlight >= this.max) {
      this.rejections++;
      return false;
    }
    this.inFlight++;
    return true;
  }

  release(): void {
    if (this.inFlight > 0) {
      this.inFlight--;
    }
  }

  snapshot(): BulkheadSnapshot {
    return {
      maxConcurrency: this.max,
      inFlight: this.inFlight,
      rejections: this.rejections,
    };
  }
}

File: curriculum/13_api_gateway_circuit_breaker/go-impl/gateway/bulkhead.go
package gateway

import (
	"sync"
)

type Bulkhead struct {
	mu         sync.Mutex
	max        int
	inFlight   int
	rejections int64
}

func NewBulkhead(max int) *Bulkhead {
	return &Bulkhead{max: max}
}

func (b *Bulkhead) Acquire() bool {
	b.mu.Lock()
	defer b.mu.Unlock()
	if b.inFlight >= b.max {
		b.rejections++
		return false
	}
	b.inFlight++
	return true
}

func (b *Bulkhead) Release() {
	b.mu.Lock()
	defer b.mu.Unlock()
	if b.inFlight > 0 {
		b.inFlight--
	}
}

func (b *Bulkhead) Snapshot() BulkheadSnapshot {
	b.mu.Lock()
	defer b.mu.Unlock()
	return BulkheadSnapshot{
		MaxConcurrency: b.max,
		InFlight:       b.inFlight,
		Rejections:     b.rejections,
	}
}

type BulkheadSnapshot struct {
	MaxConcurrency int
	InFlight       int
	Rejections     int64
}